src/wasm-lib/kcl/src/execution/kcl_value.rs

```diff
@@ -647,6 +693,78 @@ impl KclValue {
     }
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub enum RuntimeType {
+    Primitive(PrimitiveType),
+    Array(PrimitiveType),
+    Tuple(Vec<PrimitiveType>),
+    Object(Vec<(String, RuntimeType)>),
+}
+
+impl RuntimeType {
+    pub fn from_parsed(value: Type, settings: &super::MetaSettings) -> Option<Self> {
+        match value {
+            Type::Primitive(pt) => Some(RuntimeType::Primitive(PrimitiveType::from_parsed(pt, settings)?)),
+            Type::Array(pt) => Some(RuntimeType::Array(PrimitiveType::from_parsed(pt, settings)?)),
+            Type::Object { properties } => Some(RuntimeType::Object(
+                properties
+                    .into_iter()
+                    .map(|p| {
+                        p.type_.and_then(|t| {
+                            RuntimeType::from_parsed(t.inner, settings).map(|ty| (p.identifier.inner.name, ty))
+                        })
+                    })
+                    .collect::<Option<Vec<_>>>()?,
+            )),
+        }
+    }
+
+    // Subtype with no coercion, including refining numeric types.
+    fn subtype(&self, sup: &RuntimeType) -> bool {
+        use RuntimeType::*;
+
+        match (self, sup) {
+            (Primitive(t1), Primitive(t2)) | (Array(t1), Array(t2)) => t1 == t2,
```

Comment:
Because arrays are immutable, shouldn't we recurse to allow covariance?

Reply:
Probably. but I'll just leave a TODO (as for records) since for now subtyping is extremely limited and only really provides equality. Also, array types can only contain primitive types and there is no primitive subtyping yet, so it doesn't actually make any difference.